src/programacion/empresaprogramacion/Empresa.java

```diff
@@ -1,54 +1,216 @@
 package programacion.empresaprogramacion;
+
+import programacion.empresaprogramacion.excepciones.ParametroInvalidoException;
+import programacion.empresaprogramacion.excepciones.RutaArchivoInvalida;
+
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
 import java.util.Map;
+import java.util.TreeMap;
 
 public class Empresa {
-    
-    public String nombre;
-    public Map<Integer, Empleado> mapaEmpleados;
 
-    public Empresa(String nombre) {
-        this.nombre = nombre;
-        //TODO this.empleados = new ?<>();
-    }
+  public String nombre;
+  public Map<String, Empleado> mapaEmpleados;
+  static final String CSV_FILENAME_PROGRAMADORES = "programadores.csv";
+  static final String CSV_FILENAME_GERENTES = "gerentes.csv";
 
-    public void addEmpleado(Empleado empleado) {
-        // TODO  (no permitir meter 2 empleados con el mismo "idEmpleado" y dado el caso lanzar 
-        //exception)
+  public Empresa(String nombre) {
+    this.nombre = nombre;
+    mapaEmpleados = new HashMap<>();
+  }
+
+  public void addEmpleado(Empleado empleado) throws ParametroInvalidoException {
+    if (mapaEmpleados.containsKey(empleado.DNI)) {
+      if (empleado.tipoEmpleado == TipoEmpleado.PROGRAMADOR) {
+        throw new ParametroInvalidoException("Ya existe un Programador con ese ID de empleado.\n");
+      } else {
+        throw new ParametroInvalidoException("Ya existe un Gerente con ese ID de empleado.\n");
+      }
+    } else {
+      mapaEmpleados.put(empleado.DNI, empleado);
     }
+  }
+
+  public void guardarEnCSV(String archivoGuardar) throws RutaArchivoInvalida {
+    List<Programador> programadores = new ArrayList<>();
+    List<Gerente> gerentes = new ArrayList<>();
 
-    public void guardarEnCSV(String archivoGuardar) {
-        //TODO 
+    for (Empleado empleado : mapaEmpleados.values()) {
+      if(empleado instanceof Programador){
+        programadores.add((Programador) empleado);
+      } else {
+        gerentes.add((Gerente) empleado);
+      }
     }
 
-    public void cargarDesdeCSV(String archivoCargar) {
-        //TODO
+    Iterator<Programador> iteradorProgramadores = programadores.iterator();
+    Iterator<Gerente> iteradorGerentes = gerentes.iterator();
 
+    try{
+      File archivo = new File(archivoGuardar);
+      if (archivoGuardar.equalsIgnoreCase(CSV_FILENAME_PROGRAMADORES)){
+        if (comprobarArchivo(archivo)){
+          if (archivo.createNewFile()){
+            System.out.println("El archivo se ha creado correctamente.");
+          } else {
+            throw new RutaArchivoInvalida("El archivo no se ha podido crear.");
+          }
+        }
+        try (BufferedWriter bw = new BufferedWriter(new FileWriter(archivo))){
```

Comment:
Nombres de variables que indiquen lo que representan:
```suggestion
        try (BufferedWriter escritorConBuffer = new BufferedWriter(new FileWriter(archivo))){
```

Reply:
Con esto me surgen dudas. Digamos que para estas clases estilo Scanner, BufferedWriter, BufferedReader, Iterator, etc... Hay como una especie de regla no escrita (a lo mejor está escrita y no lo sé) que es llamar a ese objeto con las iniciales de la clase, es decir, Scanner = sc, BufferedReader = br, BufferedWriter = bw, Iterator = iter.... En este código mismamente llamo a los iteradores con algún otro nombre del estilo "iteradorProgramadores" para ver qué decías al respecto. En el examen, entiendo también, que como bien dices, nombres de variables significativos, ¿no?